pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java

```diff
@@ -90,32 +90,55 @@ static Pair<FieldSpec.DataType, Object> inferLiteralDataTypeAndValue(String lite
   public LiteralContext(Literal literal) {
     Preconditions.checkState(literal.getFieldValue() != null,
         "Field value cannot be null for field:" + literal.getSetField());
-    switch (literal.getSetField()){
+    switch (literal.getSetField()) {
       case BOOL_VALUE:
         _type = FieldSpec.DataType.BOOLEAN;
         _value = literal.getFieldValue();
+        _bigDecimalValue = PinotDataType.BOOLEAN.toBigDecimal(_value);
         break;
       case DOUBLE_VALUE:
-        _type = FieldSpec.DataType.DOUBLE;
-        _value = literal.getFieldValue();
+        Number floatingNumber = NumberUtils.createNumber(literal.getFieldValue().toString());
+        _bigDecimalValue = new BigDecimal(literal.getFieldValue().toString());
+        if (floatingNumber instanceof Float) {
+          _type = FieldSpec.DataType.FLOAT;
+          _value = _bigDecimalValue.floatValue();
+        } else {
+          _type = FieldSpec.DataType.DOUBLE;
+          _value = _bigDecimalValue.doubleValue();
+        }
         break;
       case LONG_VALUE:
-        _type = FieldSpec.DataType.LONG;
-        _value = literal.getFieldValue();
+        Number integralNumber = NumberUtils.createNumber(literal.getFieldValue().toString());
```

Comment:
Can be simplified with checking if the long value is between Integer.MIN to Integer.MAX